tfhe/src/c_api/high_level_api/threading.rs

```diff
@@ -0,0 +1,114 @@
+use super::super::utils::{catch_panic, get_mut_checked};
+use crate::c_api::high_level_api::keys::ServerKey;
+use crate::c_api::utils::get_ref_checked;
+use rayon::{ThreadPool, ThreadPoolBuilder};
+use std::ffi::{c_int, c_void};
+
+/// Used to limit number of threads
+///
+/// This struct can be used to limit the number of threads
+/// operations (ran inside this context) can use.
+///
+/// A threading context creates and hold the specified number of threads
+pub struct TfheThreadingContext {
+    pool: ThreadPool,
+}
+
+/// Creates a new threading context
+///
+/// - num_threads: number of threads inside this context 0 means it will number of CPU threads
+#[no_mangle]
+pub unsafe extern "C" fn tfhe_threading_context_create(
+    num_threads: usize,
+    context: *mut *mut TfheThreadingContext,
+) -> c_int {
+    catch_panic(|| {
+        *context = std::ptr::null_mut();
+
+        let pool = ThreadPoolBuilder::new()
+            .num_threads(num_threads)
+            .build()
+            .unwrap();
+
+        let ctx = TfheThreadingContext { pool };
+
+        *context = Box::into_raw(Box::new(ctx));
+    })
+}
+
+/// Destroys the threading context
+///
+/// `context` may be NULL
+#[no_mangle]
+pub unsafe extern "C" fn tfhe_threading_context_destroy(context: *mut TfheThreadingContext) {
+    let _ = catch_panic(|| {
+        if context.is_null() {
+            return;
+        }
+
+        let _ = Box::from_raw(context);
+    });
+}
+
+/// Sets the server key inside of all of the threads the context has
+#[no_mangle]
+pub unsafe extern "C" fn tfhe_threading_context_set_server_key(
+    context: *mut TfheThreadingContext,
+    server_key: *const ServerKey,
+) -> c_int {
+    catch_panic(|| {
+        let sks = get_ref_checked(server_key).map(|sks| &sks.0).unwrap();
+        let context = get_mut_checked(context).unwrap();
+
+        context
+            .pool
+            .broadcast(|_| crate::high_level_api::set_server_key(sks.clone()));
+    })
+}
+
+/// Runs the given function inside the context
+///
+///
+/// Both the `func` and `data` must be thread-safe
+#[no_mangle]
+pub unsafe extern "C" fn tfhe_threading_context_run(
+    context: *mut TfheThreadingContext,
+    func: extern "C" fn(*mut c_void) -> c_int,
+    data: *mut c_void,
+) -> c_int {
+    // *mut c_void is not Send by default
+    // neither extern fn is
+    //
+    // However rayon/rust require a `Send` closure
+    // here we can't prove it, the responsibility is on the user
+    // since the function is unsafe its fine (no soundness issue)
```

Comment:
what do you mean here no soundness issue ?

Reply:
If the user somehow breaks the thread-safety promise he made, we have undefined behaviour, but its ok as the rust function is unsafe its ok not soundness problem

Reply:
understood